### PR TITLE
Support typeof(null) mangling.

### PR DESCRIPTION
**Pull request by TurkeyMan**
Fix issue [15388 extern(C++) - typeof(null) should mangle as nullptr_t](https://issues.dlang.org/show_bug.cgi?id=15388)

`nullptr_t` didn't work. Now it does.

I don't understand the last 2 lines in the win implementation, but all the other functions had them, so I figured they were important ;)

### CODE REVIEW

**Comment by ibuclaw** [comment]
Missing testsuite entries.

**Comment by TurkeyMan** [comment]
Yeah, already on it!

**Comment by TurkeyMan** [comment]
Oh ffs, `nullptr_t` is not intrinsic on linux >_<

**Comment by TurkeyMan** [comment]
Wtf?! I give up. It's too hard!

`nullptr_t` is in `<cstddef>`, which is clearly included...
It builds for me. I don't know what it's complaining about!

**Comment by TurkeyMan** [comment]
Wtf?! I give up. It's too hard!
`nullptr_t` is in `<cstddef>`, which is clearly included...
It builds for me. I don't know what it's complaining about!

**Comment by TurkeyMan** [comment]
What C compilers are being used? Are they >7 years old?

**Comment by ibuclaw** [comment]
Having a glance before you reverted, it _looked_ like you weren't updating the right make files.  The std version used in the single test should be updated, not the std version when building dmd itself.

I wouldn't know off the top of my head which version of the abi we support, looks like nullptr is from version 7 and up.  Maybe its about time we allow adjusting the baseline from the command-line, linux, freebsd and solaris would all support differing C++ versions anyway.

**Comment by TurkeyMan** [comment]
I'll suggest there's a wider conversation here... and I think it should probably be a follow-up.
If I just add `-std=c++11` to the single test, that gets built with a bunch of compilers, and they all need to be handled/supported.
Making sure the build and test infra supports C++11 is way outside the scope of this PR.

**Comment by TurkeyMan** [comment]
So... is this okay then?

**Comment by jacob-carlborg** [comment]
We could just test the mangling with out compiling any C++ code, i.e. https://github.com/dlang/dmd/blob/master/test/compilable/cppmangle.d.

**Comment by TurkeyMan** [comment]
I did that.

File: test/compilable/cppmangle.d
Comment (jacob-carlborg) [inline test/compilable/cppmangle.d]: I recommend `Posix` if the mangling is the same.

File: test/compilable/cppmangle.d
Comment (jacob-carlborg) [inline test/compilable/cppmangle.d]: Perhaps on version statement instead, i.e. `Windows`. Or is there a particular reason to have two?

File: test/compilable/cppmangle.d
Comment (TurkeyMan) [reply test/compilable/cppmangle.d]: I used this because it appeared further above in the same file.

File: test/compilable/cppmangle.d
Comment (TurkeyMan) [reply test/compilable/cppmangle.d]: I followed the pattern from higher in the file.

File: test/compilable/cppmangle.d
Comment (jacob-carlborg) [reply test/compilable/cppmangle.d]: Ok.

File: test/compilable/cppmangle.d
Comment (jacob-carlborg) [reply test/compilable/cppmangle.d]: Ok.

Comment (TurkeyMan) [reply test/compilable/cppmangle.d]: I'm into conformity, and It's not the business of this PR to tidy someone else's code.

File: src/dmd/cppmangle.d
Comment (jacob-carlborg) [inline src/dmd/cppmangle.d]: Not sure how well `immutable`/`shared` is covered in the tests for other types but this looks like it's not covered for this type.

Comment (TurkeyMan) [reply src/dmd/cppmangle.d]: Correct. I matched the precedent.

File: test/compilable/cppmangle.d
Comment (jacob-carlborg) [reply test/compilable/cppmangle.d]: That's absolutely fine.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by TurkeyMan** [comment]
I simplified the tests.

**Comment by jacob-carlborg** [comment]
This looks fine to me. I'll let someone else have a look as well before merging.

**Comment by jacob-carlborg** [comment]
Perhaps a changelog entry as well.

**Comment by WalterBright** [comment]
> Fix issue 15388

It would be nice to use hyperlinks.

**Comment by TurkeyMan** [comment]
This is *exactly* the text Jacob told me to write... so, what should I write?

**Comment by jacob-carlborg** [comment]
> This is exactly the text Jacob told me to write... so, what should I write?

What I told was about the commit message. I think Walter is referring to the code for the tests. He likes to have a link to the issue next to the tests. He could have made that more clear by commenting on a specific line in the code.

**Comment by TurkeyMan** [comment]
Oh... yeah I don't really wanna do another amend-CI turnaround cycle :/

**Comment by jacob-carlborg** [comment]
@TurkeyMan I added the link in the tests. I can add the automerge label if you squash all commits and write the commit message so that the dlang-bot will properly link the PR and the issue, i.e `Fix issue 15388 extern(C++) - typeof(null) should mangle as nullptr_t`.

**Comment by TurkeyMan** [comment]
Won't dlang bot pull it from the merge commit's comment (ie, the text at the top of this PR)??
Do I really have to make the patch itself contain the proper text?

**Comment by TurkeyMan** [comment]
Also, out of curiosity... how did you push to my repo? O_o
I'm the only one with write access to my own repo.

**Comment by wilzbach** [comment]
> Won't dlang bot pull it from the merge commit's comment (ie, the text at the top of this PR)??

Nope. You never mention the issue in your commits. There's no way to check for the issue when latering generating the changelog.

> I'm the only one with write access to my own repo.

Maintainers have by default push access to your PR branch. It's a very handy feature.

**Comment by wilzbach** [comment]
> It's written in the description of this PR (at the top), that text is used as the commit comment for the merge commit... 

No, only the title is used.
Example: https://github.com/dlang/dmd/commit/d1ee5892181fe0c9f69e79c764d7224225ba5362

BTW the vibe.d failure looks unrelated. I have temporarily disabled it: https://github.com/dlang/ci/pull/207

**Comment by TurkeyMan** [comment]
Oh, well then I should put that text in the title then ;)

**Comment by jacob-carlborg** [comment]
>> Also, out of curiosity... how did you push to my repo? O_o
I'm the only one with write access to my own repo.

> Maintainers have by default push access to your PR branch. It's a very handy feature.

It's particularly useful when it's just one small thing remaining to bring a PR across the finish line.

**Pull request merged**
